### PR TITLE
Dropped duplicated math function from Thrust

### PR DESCRIPTION
**Pull request by viralbhadeshiya**
## Description

closes https://github.com/NVIDIA/cccl/issues/6104

- Removed duplicated math function from Thrust
- Moved all duplicated math function call from Thrust to cuda/std or cuda

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: thrust/thrust/detail/integer_math.h [hunk 1]
```diff
@@ -27,6 +27,9 @@
 #endif // no system header
 #include <thrust/detail/type_deduction.h>
 
+#include <cuda/__cmath/ceil_div.h>
+#include <cuda/__cmath/round_down.h>
+#include <cuda/__cmath/round_up.h>
```
Comment (bernhardmgruber) [inline thrust/thrust/detail/integer_math.h]: Suggestion: I don't think this is needed, since we do not use any new functions in this header file.
```suggestion
```

File: cub/cub/device/dispatch/dispatch_merge_sort.cuh
Comment (miscco) [inline cub/cub/device/dispatch/dispatch_merge_sort.cuh:44]: as @bernhardmgruber said, please remove the unused include from files

File: cub/cub/device/dispatch/dispatch_merge_sort.cuh [hunk 2]
```diff
@@ -282,7 +283,7 @@ struct DispatchMergeSort
         break;
       }
 
-      const int num_passes = static_cast<int>(THRUST_NS_QUALIFIER::detail::log2_ri(num_tiles));
+      const int num_passes = static_cast<int>(::cuda::ceil_ilog2(num_tiles));
```
Comment (miscco) [inline cub/cub/device/dispatch/dispatch_merge_sort.cuh]: The `static_cast` is not necessary, as `cuda::ceil_ilog2` returns `int`
```suggestion
      const int num_passes = ::cuda::ceil_ilog2(num_tiles);
```

File: thrust/thrust/detail/execute_with_allocator.h [hunk 3]
```diff
@@ -50,7 +52,7 @@ get_temporary_buffer(thrust::detail::execute_with_allocator<Allocator, BaseSyste
 
   // How many elements of type value_type do we need to accommodate n elements
   // of type T?
-  size_type num_elements = divide_ri(sizeof(T) * n, sizeof(value_type));
+  size_type num_elements = ::cuda::ceil_div(sizeof(T) * n, sizeof(value_type));
```
Comment (miscco) [inline thrust/thrust/detail/execute_with_allocator.h]: pre-existing: This might lead to conversion warnings, depending on what type `size_type` denotes. We should static_cast
```suggestion
  const size_type num_elements = static_cast<size_type>(::cuda::ceil_div(sizeof(T) * n, sizeof(value_type)));
```

Applies throughout

File: thrust/thrust/mr/disjoint_pool.h [hunk 4]
```diff
@@ -370,7 +372,7 @@ class disjoint_unsynchronized_pool_resource final
   [[nodiscard]] void_ptr do_allocate_impl(std::size_t bytes, std::size_t alignment)
   {
     bytes = (std::max) (bytes, m_options.smallest_block_size);
-    assert(detail::is_power_of_2(alignment));
+    assert(::cuda::ceil_ilog2(alignment));
```
Comment (miscco) [inline thrust/thrust/mr/disjoint_pool.h]: Critical: This is not correct
```suggestion
    assert(::cuda::is_power_of_two(alignment));
```

File: thrust/thrust/detail/integer_math.h (hunk 1 above)
Comment (viralbhadeshiya) [reply thrust/thrust/detail/integer_math.h]: Fixed.

File: cub/cub/device/dispatch/dispatch_merge_sort.cuh
Comment (viralbhadeshiya) [reply cub/cub/device/dispatch/dispatch_merge_sort.cuh:44]: Fixed.

File: thrust/thrust/detail/execute_with_allocator.h (hunk 3 above)
Comment (viralbhadeshiya) [reply thrust/thrust/detail/execute_with_allocator.h]: Fixed.

File: thrust/thrust/mr/disjoint_pool.h (hunk 4 above)
Comment (viralbhadeshiya) [reply thrust/thrust/mr/disjoint_pool.h]: Fixed.

**Comment by viralbhadeshiya** [comment]
@bernhardmgruber @miscco all changes that you guys suggested is done, lmk if there are more Changes. :)

**Comment by miscco** [comment]
@viralbhadeshiya This looks great 🎉 

Thanks a lot for working on this.

I added a few minor changes:
* dropped unused `is_odd`
* improved `is_negative`
* dropped two more includes

File: thrust/thrust/detail/integer_math.h [hunk 5]
```diff
@@ -25,89 +25,29 @@
...
-  if (!is_power_of_2(x))
+  if constexpr (::cuda::std::is_unsigned_v<T>)
   {
-    ++result;
+    return false;
+  }
+  else
+  {
+    return x < 0;
   }
```
Comment (bernhardmgruber) [inline thrust/thrust/detail/integer_math.h]: Suggestion: we can simplify this even further:
```c++
return ::cuda::std::is_signed_v<T> && x < 0;
```

Comment (miscco) [reply thrust/thrust/detail/integer_math.h]: That is not a simplification, because the compiler most often cannot elide the `::cuda::std::is_signed_v<T>`

The if constexpr is the most efficient way

Comment (bernhardmgruber) [reply thrust/thrust/detail/integer_math.h]: Wat, really? The compiler just has to fold a constant. But I believe you if you say so.

**Comment by viralbhadeshiya** [comment]
Seems like couple of jobs are failing, @miscco lmk if it requires any changes for it, happy to do it.

File: thrust/thrust/detail/integer_math.h [hunk 6]
```diff
@@ -25,89 +25,29 @@
...
-_CCCL_HOST_DEVICE _CCCL_FORCEINLINE bool is_odd(Integer x)
-{
-  return 1 & x;
-}
-
-template <typename Integer>
-_CCCL_HOST_DEVICE _CCCL_FORCEINLINE Integer log2(Integer x)
+_CCCL_TEMPLATE(typename T)
+_CCCL_REQUIRES(::cuda::std::is_arithmetic_v<T>)
+_CCCL_HOST_DEVICE _CCCL_FORCEINLINE constexpr bool is_negative([[maybe_unused]] const T x) noexcept
```
Comment (fbusato) [inline thrust/thrust/detail/integer_math.h]:
```suggestion
[[nodiscard]] _CCCL_API constexpr bool is_negative([[maybe_unused]] T x) noexcept
```

File: thrust/thrust/detail/integer_math.h [hunk 7]
```diff
@@ -25,89 +25,29 @@
 #elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
 #  pragma system_header
```
Comment (fbusato) [inline thrust/thrust/detail/integer_math.h]: would make sense to also update the license because the file is practically entirely new.

Comment (viralbhadeshiya) [reply thrust/thrust/detail/integer_math.h]: @fbusato how can I do that?

Comment (miscco) [reply thrust/thrust/detail/integer_math.h]: Honestly, this is effectively only ever used in `thrust/system/cuda/detail/dispatch.h`, so lets move the function there and drop the file?

Comment (viralbhadeshiya) [reply thrust/thrust/detail/integer_math.h]: @miscco I see couple of usage of ```is_negative``` in ```thrust/thrust/system/cuda/detail/scan.h``` too, do we need to worry about that?

Comment (miscco) [reply thrust/thrust/detail/integer_math.h]: It is already including `dispatch.h`

**Review by bernhardmgruber — APPROVED** [review]

**Comment by miscco** [comment]
@fbusato There is an issue with `thrust::pool_options`

I suspect there is a differing behavior for `is_power_of_two` could you please inevestigate why it differs?

**Comment by fbusato** [comment]
`::cuda::is_power_of_two` correctly evaluates 0 as non-power of two, while `thrust::detail::is_power_of_2()` misses the check `x != 0`.

This causes a different behavior in `!::cuda::is_power_of_two(alignment)`, which is initialized to 0 by default.
https://github.com/NVIDIA/cccl/pull/6188/files#diff-d85bf89d70ec2f2a069c4c0ffeccd2722fe2a108a012566503392885900f40a1R115

**Comment by fbusato** [comment]
@viralbhadeshiya my suggestion is just to add the condition `alignment != 0 && !::cuda::is_power_of_two(alignment)`

**Comment by viralbhadeshiya** [comment]
> `::cuda::is_power_of_two` correctly evaluates 0 as non-power of two, while `thrust::detail::is_power_of_2()` misses the check `x != 0`.
> 
> This causes a different behavior in `!::cuda::is_power_of_two(alignment)`, which is initialized to 0 by default. https://github.com/NVIDIA/cccl/pull/6188/files#diff-d85bf89d70ec2f2a069c4c0ffeccd2722fe2a108a012566503392885900f40a1R115

@fbusato would this similar changes also required in ```!::cuda::is_power_of_two(smallest_block_size)``` & ```!::cuda::is_power_of_two(largest_block_size)```?

**Comment by fbusato** [comment]
yes, but I'm not sure that this edge case is evaluated in the test case. You can also run the test locally to make sure it works

**Comment by viralbhadeshiya** [comment]
> yes, but I'm not sure that this edge case is evaluated in the test case. You can also run the test locally to make sure it works

Okay made changes, unfortunately my remote server which had NVIDIA device is facing some issue so could not run tests. but changes are done.
